wandb/client
# Add pytest shard for notebook testing

Description
-----------
Shard pytest unit tests into finer grained groupings

Why?
- performance: requirements_dev.txt is getting large, the size of this file impacts the performance of running tests because these requirements have to be downloaded and built (or retrieved from cache).  circle doesnt behave well with caches larger than 500MB
- organization: eventually sharding into different directories should make it easier to know where to add your test and limit the scope of what you need to understand when debugging
- customizability: by sharding we can enable some features for certain shards, for example multiprocessing coverage collection
- metrics: finer grained metrics about coverage with sharded pytest -- eventually

Why this might be a bad idea
- more complicated,  as you add a new test, you might need to think harder about where it goes, what resources it uses -- this can be mitigated with better devel docs
- complexity in the short term -- codecov counts have to be manually increased for each shard that is collecting converage
- In transition between non sharded to sharded tests there might be some code duplication (config files for testing)

Near term future work:
- move more things into shards (tensorflow / pytorch?)
- conftest.py cleanups to not load everything (split out fixtures into separate files /shards?)
- enable multiprocessing coverage in the notebook shard to get more coverage credit for notebook tests (might be harder than I realized -- consider switching to: https://github.com/chrisjsewell/pytest-notebook/pull/3/commits/5e9659c25dd16fbb52313343c31fd912e9acb6ba -- more details about how this works: https://github.com/chrisjsewell/pytest-notebook/blob/40381be0963a8866d1a46994595844bbfb4e66c1/pytest_notebook/execution.py#L77 -- note, it injects cells to setup and teardown coverage.py...)
- add some make targets to support recommended testing for different types of platforms /use cases.   right now we have `make test-short`.   update dev docs based on these recommended top level commands.  for example  if your change has nothing to do with ml frameworks you might not want to wait for all that to run.

Testing
-------
- Run the circleci suite and make sure that coverage is similar.
